### ISSUE TITLE
Tracker bug for adding Msbuild support

### ISSUE THREAD

**Issue by tmds** [redhat-developer/s2i-dotnetcore#46]
Track info and activities on changes needed for change from project.json to msbuild/csproj.

Tasks:

- [ ] detect default assembly name - `project.json` defaults to the `<parentfolder>.dll`. `<name>.csproj` defaults to `<name>.dll`
- [ ] restore handling - `dotnet restore` will no longer work recursively, but it will work transitive on the referenced projects. We can deprecate `DOTNET_RESTORE_ROOT` and restore `DOTNET_STARTUP_PROJECT` and `DOTNET_TEST_PROJECTS`.

/cc @omajid @jerboaa

**Comment by tmds** [comment]
@jerboaa @omajid We can consider removing `DOTNET_RESTORE_ROOT` already. Using `DOTNET_RESTORE_ROOT` a user can trim the projects to be restored which may significantly reduce the restore time and he can eliminate projects which are causing restore issues. For me, these use-cases are no longer compelling to introduce a new envvar which will be deprecated by msbuild. What do you think? Keep or remove?

**Comment by jerboaa** [comment]
I don't expect apps to move to msbuild/csproj this quickly. So for a transition time it makes sense to have good support for both. That's my $0.02.

**Comment by tmds** [comment]
> I don't expect apps to move to msbuild/csproj this quickly. So for a transition time it makes sense to have good support for both. That's my $0.02.

afaik the cli won't support both simultaneously.

**Comment by jerboaa** [comment]
Assuming you mean .NET Core CLI, this is something part of the RPMs. So it sounds like until we ship updates to RPMs and then an updated image there is no way to add csproj support, right? Seems to be a 2.0 target in that case. We should investigate whether cli actually will support both or are mutually exclusive. We need a definitive answer in order to plan correctly.

**Comment by tmds** [comment]
> Assuming you mean .NET Core CLI, this is something part of the RPMs. So it sounds like until we ship updates to RPMs and then an updated image there is no way to add csproj support, right? Seems to be a 2.0 target in that case.

Correct.

> We should investigate whether cli actually will support both or are mutually exclusive. We need a definitive answer in order to plan correctly.

They are mutually exclusive. project.json support is dropped.

**Comment by omajid** [comment]
Sorry if this is obvious, but how will msbuild work with `DOTNET_RESTORE_ROOT`? Will `dotnet restore /some/path` continue working? How are the use-cases that `DOTNET_RESTORE_ROOT` was helping with for `project.json` different for `csproj`?

And Microsoft is pushing `csproj` pretty hard. Their docker container tags for already-released 1.1 for example, are going to switch from `project.json` to `csproj`.

**Comment by tmds** [comment]
> Sorry if this is obvious, but how will msbuild work with DOTNET_RESTORE_ROOT? Will dotnet restore /some/path continue working? How are the use-cases that DOTNET_RESTORE_ROOT was helping with for project.json different for csproj?

msbuild will handle the relationships between the projects. So if project app1 uses project lib1, then restoring app1 will also restore lib1. With project.json both had to be restored.
The restore command under project.json did work recursively. So restoring . at the root found all project.json and would restore app1 and lib1. msbuild will no longer restore recursively.

So when we move to project.json, we will restore DOTNET_STARTUP_PROJECT and DOTNET_TEST_PROJECTS. The magic of MsBuild will figure out the dependent projects to be restored. This makes it no longer necessary to be able to extend/trim projects via DOTNET_RESTORE_ROOT.

If we remove DOTNET_RESTORE_ROOT for msbuild, we may as well remove it already and default to '.'. This has the downside of occasionally restoring to many projects. The pro is we won't introduce a new variable which will be deprecated in a couple of months.

**Comment by omajid** [comment]
> This has the downside of occasionally restoring to many projects.

This con sounds acceptable to me: I can live with a slower container build time, especially if it's not a regression from the already-released version.

Based on my understanding, I am okay with reverting DOTNET_RESTORE_ROOT support.

**Comment by tlbdk** [comment]
How is it going with this one? With .NET Core SDK 1.0.1 (Runtime 1.1.1) and VS2017 only sln and csproj files are supported.

**Comment by jerboaa** [comment]
@tlbdk My understanding is that csproj support will come with 2.0 .NET Core containers. It's subject to change, though. The recommendation is to stick with older Visual Studio for current images for now.

**Comment by tlbdk** [comment]
@jerboaa Sadly it's not the case, the first stable version of .net core tooling only supports csproj and this is the version VS 2017 is using. VS 2017 will also auto-upgrade the project.json to csproj files without asking.

Effectively this means that project.json is more of less dead unless you are limited to using an old preview version of the tooling and SDK.

**Comment by tlbdk** [comment]
Also non of the new runtime releases (1.1.1 and 1.0.4) are done with the old preview tool that supported project.json: 
https://github.com/dotnet/core/blob/master/release-notes/download-archive.md

**Comment by tmds** [comment]
@tlbdk our recommendation is to keep using VS2015 for now.
We have a strong desire to stay close with the upstream but in this case this is a challenge.

Since we are packaging the SDK and runtime together, we could not ship a backwards incompatible CLI as part of a patch release (1.1.0 -> 1.1.1). We are looking at how we should improve our packaging.

Next, the CLI has a dependency on the 1.1 runtime. This is the case even when developing for 1.0. In our current packaging, 1.0 and 1.1 are separate.

These two things, mean that we would have introduce a new set of packages and containers for 1.x. Building 1.x from source is a challenge and this is adding to that challenge. We are actively working with Microsoft on improving this for 2.0.

We had identified the VS2017 issue and our current plan is to invest our efforts in 2.0 which will include msbuild and is due in the summer.

> Also non of the new runtime releases (1.1.1 and 1.0.4) are done with the old preview tool that supported project.json:

We have updated our packages and containers to include the new runtime (with the old CLI). On Windows, the runtimes and sdks can be installed side-by-side. Using global.json you can specify the desired sdk.

**Comment by tlbdk** [comment]
I understand your concerns, but it leaves us in a bit of a pinch for using openshift for .net core, the 2.0 release seems to be scheduled for Q3(if that holds), our deadline is quite a bit before:
https://github.com/dotnet/core/blob/master/roadmap.md

Having to use the old SDK for some projects and/or upgrade the runtime but not the tooling(is there a guide for that?) is also problematic at best, we have developer on window, mac and linux.

Also not being able to use VS 2017 even though .net core support is limited on VS2015 is also a bit of a dealbreak at the moment as this seems to be the most stable IDE for .net core. We are also using Rider and VSCode on both Mac and Linux and both IDE's support csproj and the seem to be moving away from project.json.

Could we do something where we have both tooling installed at the same time and then do a runtime selection on what we are using, so we don't break backward or forwards compatibility?

**Comment by tmds** [comment]
>  understand your concerns, but it leaves us in a bit of a pinch for using openshift for .net core, the 2.0 release seems to be scheduled for Q3(if that holds), our deadline is quite a bit before:

The limitation is not being able to use VS2017/msbuild. Deployed applications get an updated runtime.

> Having to use the old SDK for some projects and/or upgrade the runtime but not the tooling(is there a guide for that?) is also problematic at best, we have developer on window, mac and linux.

The simplest solution is to not install the msbuild SDK. When you deploy, it will use the newer runtime provided in the container. If you are hitting one of the bugs fixed in the patch release during development, this is not an option. If you want to use VS2017, this is not an option either.

Then you can install multiple sdks side-by-side.
In the solution root for the OpenShift projects, you can select the project.json sdk via global.json:
```
{
    "sdk": {
        "version": "1.0.0-preview2-1-003177"
    }
}
```
The project.json sdks are the ones labeled preview2.

> Also not being able to use VS 2017 even though .net core support is limited on VS2015 is also a bit of a dealbreak at the moment as this seems to be the most stable IDE for .net core. We are also using Rider and VSCode on both Mac and Linux and both IDE's support csproj and the seem to be moving away from project.json.

Microsoft is strongly pushing msbuild by having no support for project.json in VS2017. I haven't used VS2017, so I don't know what improved compared to VS2015. Rider support for csproj has just landed, so I think project.json may be actually working better there. Everyone will move away from project.json to msbuild. Our current plan is to do it as part of 2.0.

> Could we do something where we have both tooling installed at the same time and then do a runtime selection on what we are using, so we don't break backward or forwards compatibility?

During development, the global.json sdk allows this selection. The published application will use the runtime provided by the container. Note that even when your app is specified to use 1.1.0, it will automatically use 1.1.1 if that is available. To publish your app, you need to set the newly added DOTNET_PUBLISH envvar to true.

**Comment by tlbdk** [comment]
I'm looking at the build part using s2i and I already upgraded all my projects to use the new csproj format because of issues with tooling for the old SDK's, fx. Rider could also not open the projects with the old format, some of the code will also be built on TFS and installing the old SDK's are not an option. At least for us this means s2i will not be an option until fall :(

Just to understand the details, is the old SDK tooling being patched into the rpm that is being used by the container s2i-dotnetcore or is it done in this project?

Why not just package both the new sdk tooling and the old in the same image and make it possible to switch with an environment variable? It's not the first time an upstream has broken backwards compatibility and normally this has been handled, why wait in this case?

**Comment by tmds** [comment]
> I'm looking at the build part using s2i and I already upgraded all my projects to use the new csproj format because of issues with tooling for the old SDK's, fx. Rider could also not open the projects with the old format, some of the code will also be built on TFS and installing the old SDK's are not an option.

I am not familiar with the entire setup. As long as there are no projects which are built using the new tooling and the old tooling you could work around the issue. E.g. TFS produces NuGet packages which are consumed by the s2i build. It's scary to see a project system that was used for 2 years become unsupported less than 3 weeks after a replacement becomes available.

> Just to understand the details, is the old SDK tooling being patched into the rpm that is being used by the container s2i-dotnetcore or is it done in this project?

Everything is built from source. The 1.1.1 build is the 1.1.0 build with an updated framework and runtime. 

The 1.1.1 container is capable of running the new CLI but we do not ship it since we cannot build it ourselves. We are actively working on this for 2.0.

> Why not just package both the new sdk tooling and the old in the same image and make it possible to switch with an environment variable? It's not the first time an upstream has broken backwards compatibility and normally this has been handled, why wait in this case?

The switch would have to be handled in the container and bypass the dotnet muxer. If we would do a side-by-side install (cfr Windows) it would break all existing s2i builds (unless they have a global.json with an sdk property).
We also have the issue of not being able to build the cli ourselves. This is important to be able to provide support.

> At least for us this means s2i will not be an option until fall :(

A less supported option (until 2.0) would be for you to create a builder image based on the existing 1.1.1 image with a new SDK. This involves getting the sdk installer (e.g. [dotnet-install.sh](https://raw.githubusercontent.com/dotnet/cli/rel/1.0.1/scripts/obtain/dotnet-install.sh) -i <tempfolder>) and copying it's sdk at sdk/<sdk-version> to /opt/rh/dh-dotnetcore11/root/usr/lib64/dotnetcore/sdk. And performing the two changes mentioned in this issue to the assemble script.

It is not our plan to build an official 1.1.1 with csproj, but plans can change based on customer requests. This package can't ship until we can fully build it ourselves from source to be able to provide appropriate support.

**Comment by tmds** [comment]
@tlbdk does one of the options work for you until we get 2.0?

**Comment by mykas** [comment]
Hi all, we have the same issue regarding openshit and dotnetcore docker image not suppprting new cli. We were able to install sdk 1.1.1. from source but it seems there is an issue with sc enable. As I understand we need to build software collection to support it? Can someone at least help building software collection?

**Comment by tlbdk** [comment]
@tdms Sorry for the delay in getting back, I think all of them would work, but would require a bit of work from our OpenShift team, so right now we are going with deploying on our old windows setup, I hope they have time to do one of the options in the close future.

**Comment by tmds** [comment]
@mykas I have something on this branch: https://github.com/tmds/s2i-dotnetcore/tree/patchcli. This is the runtimes built by Red Hat with an sdk built by Microsoft. Executing the ./build.sh script will build an s2i builder locally. I migrated one of the tests to csproj and it succeeds. I have not tried importing the builder in OpenShift.

**Comment by dnguyen9** [comment]
> A less supported option (until 2.0) would be for you to create a builder image based on the existing 1.1.1 image with a new SDK. This involves getting the sdk installer (e.g. dotnet-install.sh -i ) and copying it's sdk at sdk/ to /opt/rh/dh-dotnetcore11/root/usr/lib64/dotnetcore/sdk. And performing the two changes mentioned in this issue to the assemble script.

@tmds Are you able to provide clarity on what you mean by "less supported"? 

> I have something on this branch: https://github.com/tmds/s2i-dotnetcore/tree/patchcli. This is the runtimes built by Red Hat with an sdk built by Microsoft. Executing the ./build.sh script will build an s2i builder locally. I migrated one of the tests to csproj and it succeeds. I have not tried importing the builder in OpenShift.

This worked for me. out of the above two options, what do you recommend as the better alternative?

**Comment by tmds** [comment]
> This worked for me. out of the above two options, what do you recommend as the better alternative?

They are the same option. The patchcli branch applies what is described in text.

> "less supported"

This is a custom builder with parts from a third-party. If you have an issue which is reproducible on the official builder, you can expect support. (Note that while running the application only bits of the official builder are used. The SDK is used during the build process).

The official builder is maintained by Red Hat. You are responsible for updating the custom builder.
You can see when the official builder is updated at https://access.redhat.com/containers/#/registry.access.redhat.com/dotnet/dotnetcore-11-rhel7/images.

A supported builder for csproj will come with the 2.0 release (see Issue 5 at https://access.redhat.com/documentation/en-us/net_core/1.1/html/release_notes/known_issues).

**Comment by dnguyen9** [comment]
> A supported builder for csproj will come with the 2.0 release (see Issue 5 at https://access.redhat.com/documentation/en-us/net_core/1.1/html/release_notes/known_issues).

Tom, will you backport the support of csproj into 1.0 & 1.1 with the 2.0 release?

**Comment by tmds** [comment]
> Tom, will you backport the support of csproj into 1.0 & 1.1 with the 2.0 release?

We have no plans to do a 1.x release with csproj support because 2.0 is a superset of 1.x. There is no compelling reason to stay at 1.x once 2.0 is available.

**Comment by cmendible** [comment]
Hmmmm this is a stopper for us. We've migrated all our projects to the new csproj msbuild tooling because project.json is "dead". We'll have to keep deploying on our old window setup till 2.0 becomes available?

Why can't the image contain both versions of the SDK side by side?

Thanks...

**Comment by tmds** [comment]
@cmendible You'll find the answers to your questions in these comments: https://github.com/redhat-developer/s2i-dotnetcore/issues/46#issuecomment-289662976, https://github.com/redhat-developer/s2i-dotnetcore/issues/46#issuecomment-292158240, https://github.com/redhat-developer/s2i-dotnetcore/issues/46#issuecomment-297906760.

**Comment by tmds** [comment]
This has been implemented on the 2.0-dev branch (https://github.com/redhat-developer/s2i-dotnetcore/pull/70).
